Add hack/update-godep-licenses.sh to generate Godeps/LICENSES.md

karlkfi opened the pull request:
Needed to get a list of dependency licenses. So I...
- Wrote an external dockerized tool (https://github.com/mesosphere/godep-licenses) to get them from Godep
- Added `Godep/LICENSES.md` with a table of packages and licenses
- Added `hack/update-godep-licenses.sh` to update `Godep/LICENSES.md`
- Added `hack/verify-godep-licenses.sh` to run `hack/update-godep-licenses.sh` and compare output against `Godep/LICENSES.md`
- Updated `make verify` to run `hack/verify-godep-licenses.sh`
- Added `hack/verify-godep-licenses.sh` to the pre-commit hook (only when `/Godeps/` has changes)

Exemptions (licenses missing or unable to parse):
- github.com/abbot/go-http-auth:Apache-2
- github.com/beorn7/perks/quantile:MIT?
- github.com/daviddengcn/go-colortext:BSD?
- github.com/shurcooL/sanitized_anchor_name:MIT?
- github.com/spf13/cobra:Apache-2
- github.com/stretchr/objx:MIT?
- github.com/stretchr/testify:MIT?

karlkfi commented:
Apparently "LICENCE" is valid UK spelling. The testify author doesn't want to change their file name. So we'll need to allow these permutations.

jdef commented:
any reason that the binary dependencies can't be captured in a docker image?

could we run the script with something like: `docker run -v $(pwd):/kubernetes kubernetes/licensing` and have the image update the licensing MD file?

jdef commented on hack/update-generated-licenses.sh:
```diff
@@ -0,0 +1,162 @@
+[ -z "${NINKA}" ] && echo "No NINKA found in environment. Please define the path to ninka.pl." && exit 1
+echo "Using Ninka: ${NINKA}"
+
+LICENSE_FILES="LICENSE LICENSE.txt LICENSE.md LICENCE.txt" # LICENSE.txt for testify - https://github.com/stretchr/testify/issues/251
```
should we add `LICENCE.md` for completeness?

jdef replied:
as well as `LICENCE`?

jdef commented on hack/update-generated-licenses.sh:
```diff
@@ -0,0 +1,162 @@
+  #local LICENSE="$(license "${KUBE_ROOT}/${ROOT_PKG_LICENSE_FILE}")"
+  #echo "${ROOT_PKG},${ROOT_PKG_LICENSE_FILE},${LICENSE}"
+  
+  cat "${KUBE_ROOT}/Godeps/Godeps.json" | jq -r ".Deps[].ImportPath" | sort -f |
```
is `jq` already a binary dep of k8s?

karlkfi replied:
nope. just used by a couple docs, afaik

jdef commented on hack/update-generated-licenses.sh:
```diff
@@ -0,0 +1,162 @@
+  cat "${KUBE_ROOT}/Godeps/Godeps.json" | jq -r ".Deps[].ImportPath" | sort -f |
+    while read PACKAGE; do
+      local P="${PACKAGE}"
+      #echo "${P}"
```
remove debug code?

sttts commented on hack/update-generated-licenses.sh:
```diff
@@ -0,0 +1,162 @@
+function pkg_license() {
+  local PREV_LICENSE_FILE=""
+  pkg_file_license | while read LINE; do
+    local PKG="$(echo "${LINE}" | sed 's/\(.*\),.*,.*/\1/')"
```
wth is this regex doing? impossible to read.

sttts replied:
there is cut and awk to filter by column.

karlkfi commented:
> any reason that the binary dependencies can't be captured in a docker image?

great idea. there's actually a dockerfile in the ninka repo, but it doesn't look to actually be hosted anywhere.

I could alternatively pull this all out to a different repo and make a dockerized tool that generates a report for any Godep-using project.

jdef commented:
let's not aim for the stars in this PR -- interested in getting this
initial cut approved and merged upstream. we could follow up in another PR
with a dockerized version. i don't think we need to solve the problem of
doing this for every Godep project.

On Tue, Dec 8, 2015 at 12:44 PM, Karl Isenberg <notifications@github.com>
wrote:

> any reason that the binary dependencies can't be captured in a docker
> image?
>
> great idea. there's actually a dockerfile in the ninka repo, but it
> doesn't look to actually be hosted anywhere.
>
> I could alternatively pull this all out to a different repo and make a
> dockerized tool that generates a report for any Godep-using project.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/18322#issuecomment-162957933>
> .
>

karlkfi commented:
Too late: https://github.com/mesosphere/godep-licenses
Stars already reached.

Commit:
Add hack/update-godep-licenses.sh to generate Godeps/LICENSES.md

- Add Godeps/LICENSES.md
- Add verify-godep-licenses to verify that Godeps/LICENSES.md is up to date
- Trigger verify-godep-licenses in the pre-commit hook only if the Godeps dir has changed
- Exclude verify-godep-licenses in verify-all
- Add verify-godep-licenses to make verify (used by travis)
- Add verify-godep-licenses to shippable
- Update dev docs to mention update-godep-licenses

karlkfi commented:
Looks like `hack/verify-godeps.sh` would fail on master, except it has an exception case to not run there. So it fails if you try to change anything in the Godeps dir.

davidopp commented:
Not trying to stall you, but @lavalamp is the best person to review this.

Maybe you can add a comment to the script about
* when should a person run this
* when should a person add/remove a `-e`

karlkfi commented:
@davidopp @lavalamp This isn't blocking us, just a nice to have to make it reproducible (now that we have the information). 

Ideally it would run automatically in the pre-commit hook every time the Godep directory is changed, but I don't know how to do that...

I can add some more comments on usage tho.

lavalamp commented on hack/update-godep-licenses.sh:
```diff
@@ -0,0 +1,43 @@
+
+cd "${KUBE_ROOT}"
+
+exec docker run --rm -i -v "${KUBE_ROOT}:/repo" mesosphere/godep-licenses:latest -p /repo \
```
Is the 'exec' necessary?

karlkfi replied:
The exec hands off process control to the executed command. This is the easiest way to propagate signal handling. Without exec it's like a dozen lines of code to properly handle SIGTERM.

lavalamp replied:
TIL

lavalamp commented on hack/update-godep-licenses.sh:
```diff
@@ -0,0 +1,43 @@
+  -e github.com/spf13/cobra:Apache-2 \
+  -e github.com/stretchr/objx:MIT? \
+  -e github.com/stretchr/testify:MIT? \
+  -o md > Godeps/LICENSES.md
```
Do you want to run it again for third_party?

karlkfi replied:
If we need one for third_party I would make it a different script and a different output file, like `third_party/LICENSES.md`.

karlkfi replied:
godep-licenses also doesn't work on generic directories. It only works on Godep vendored files. So you'd need some other similar method for trolling through a directory looking for LICENSE files, and some way to know where they came from and which ones are root repo directories.

lavalamp replied:
Gotcha. There's not that much in third_party.

lavalamp commented:
@karlkfi I would recommend adding a verification script that just checks that there's an entry for every package. Otherwise I expect this will quickly get out of sync.

karlkfi commented:
@lavalamp You mean verify the output to ensure the exemptions are up to date? I suppose we could do that, but we're already not triggering it automatically. So it's going to get out of date anyway...

I could add a mode to godep-licenses that errors when a license or exemption is not found. Is that sufficient?

lavalamp commented:
Yeah, I was meaning a script to be run by CI that at least verifies that there's an entry for everything (and not extra entries). Otherwise I don't think people will remember to run your update script.

karlkfi commented:
Ah, well that's a possibility, but I think the more consistent way to do it would be to have a pre-commit validation hook that runs this script and checks for differences whenever the Godeps directory changes. Then it would fail any time `Godeps/LICENSES.md` is out of date, and it would fail any time the generation script fails (because of a new flag that requires all packages to have licenses or exemptions).

karlkfi commented:
Unfortunately, I don't think adding those additional pieces are high priority for us atm. Do you want this script without those pieces or should I just leave this PR hanging until someone has bandwidth to add them?

lavalamp commented:
I think we can take this without the verification script. The precommit hook is nice but not everyone runs them.

However, if you look at the travis output, this actually makes the godep restore test fail, so you will either need to put this file somewhere else or make that test tolerate its presence.

karlkfi commented:
`hack/verify-godeps.sh` is failing for other reasons. It only looks for diffs in `Godeps/Godeps.json` and `Godeps/_workspace/src`.

I did figure out how to add `hack/update-godep-licenses.sh` into `hack/verify-godeps.sh`, but it's failing because master:head is dirty. It looks like because the validation never runs on master, the first person to change any Godeps after a repo is borked has to deal with the issue or not be able to commit the changes, even if it has nothing to do with their changes.

karlkfi commented:
@k8s-bot test this please

karlkfi commented:
for posterity:

```
hack/verify-godeps.sh master
Starting to download all kubernetes godeps. This takes a while
godep: restore, during download dep "github.com/docker/docker/pkg/timeutils": exit status 1
godep: restore, during download dep "github.com/docker/docker/pkg/units": exit status 1
godep: restore, during download dep "github.com/docker/libcontainer": exit status 1
godep: restore, during download dep "github.com/google/cadvisor/api": exit status 1
godep: restore, during download dep "github.com/google/cadvisor/http": exit status 1
godep: restore, during download dep "github.com/google/cadvisor/manager": exit status 1
godep: restore, during download dep "github.com/google/cadvisor/pages": exit status 1
godep: restore, during download dep "github.com/google/cadvisor/validate": exit status 1
!!! Error in hack/verify-godeps.sh:74
  '"${GODEP}" restore' exited with status 1
Call stack:
  1: hack/verify-godeps.sh:74 main(...)
Exiting with status 1
```

karlkfi commented:
forgot i could just hit the re-run button in travis...

karlkfi commented:
Looks like it was just network issues.

karlkfi commented:
Turns out adding a verify script and pre-commit hook was easier than I thought. I've also added it to travis and shippable.

karlkfi commented:
rebased & squashed

karlkfi commented:
New verify step failed on travis. Seems it doesn't have docker...

```
hack/verify-godep-licenses.sh master
Created workspace: /home/travis/gopath/src/github.com/kubernetes/kubernetes/kube-godep-licenses.HNzylH
/home/travis/gopath/src/github.com/kubernetes/kubernetes/hack/update-godep-licenses.sh: line 35: exec: docker: not found
Removing workspace: /home/travis/gopath/src/github.com/kubernetes/kubernetes/kube-godep-licenses.HNzylH
make: *** [verify] Error 127
```

Commit:
Add docker to travis builds

- Run travis builds with sudo to allow docker usage
- Run travis on trusty (for docker beta)
- https://github.com/travis-ci/travis-ci/issues/5448

karlkfi commented:
Added docker & sudo to travis.

karlkfi commented:
travis seems to be having trouble with docker not being in some build images. I logged a bug: https://github.com/travis-ci/travis-ci/issues/5448

Retrying to see it I can get an agent with docker on it by luck...

karlkfi commented:
They say it's a bug they've fixed in their "edge" group that hasn't rolled out to the "stable" group yet. So I'm adding a new commit to use edge temporarily.

karlkfi commented:
using trusty + edge seems to have worked around the travis bug.

Looks like the license validation failed properly (I must not have re-run it after rebasing).

karlkfi commented:
Is travis running on a merge branch? Guess it wants another rebase...

karlkfi commented:
oi... rebase didn't have any changes... means the script is having different output in travis...

```
Your godep licenses file is out of date. Run hack/update-godep-licenses.sh and commit the results.
--- /home/travis/gopath/src/github.com/kubernetes/kubernetes/Godeps/LICENSES.md	2016-01-13 20:07:00.125909367 +0000
+++ /home/travis/gopath/src/github.com/kubernetes/kubernetes/kube-godep-licenses.4GZBXJ/Godeps/LICENSES.md	2016-01-13 20:24:23.090072192 +0000
@@ -34,7 +34,7 @@ github.com/elazarl/goproxy | BSDWarr
 github.com/emicklei/go-restful | MITname
 github.com/evanphx/json-patch | BSDWarr
 github.com/fsouza/go-dockerclient | spdxBSD2
-github.com/garyburd/redigo/internal | ApachesPermLim
+github.com/garyburd/redigo/internal/redistest | ApachesPermLim
 github.com/garyburd/redigo/redis | ApachesPermLim
 github.com/ghodss/yaml | MITname
 github.com/go-ini/ini | Apache-2
@@ -55,7 +55,7 @@ github.com/jmespath/go-jmespath | Apache
 github.com/jonboulle/clockwork | Apache-2
 github.com/juju/ratelimit | LesserExceptionGPLVer3-TOOLONG
 github.com/kardianos/osext | spdxBSD3
-github.com/kr/pty | spdxMIT
+github.com/kr/pty | UNKNOWN
 github.com/matttproud/golang_protobuf_extensions | Apache-2
 github.com/mesos/mesos-go | Apache-2
 github.com/miekg/dns | spdxBSD3
Removing workspace: /home/travis/gopath/src/github.com/kubernetes/kubernetes/kube-godep-licenses.4GZBXJ
make: *** [verify] Error 1
```

karlkfi commented:
`github.com/kr/pty/License` might not be found because the filesystem is case-sensitive...

`github.com/garyburd/redigo/internal` -> `github.com/garyburd/redigo/internal/redistest` is weird. It looks like it might have recursed into the directories first (instead of purely alphabetically) to find the license comment in `internal/redistest/testdb.go` before the one in `internal/commandinfo.go`.

Might have to fix those in godep-licenses itself.

karlkfi commented:
Pushed an update to mesosphere/godep-licenses that fixes case-sensitivity.

karlkfi commented:
Looks like the integration test flaked:

```
F0113 20:01:28.562773    1551 controller.go:84] Unable to perform initial IP allocation check: unable to persist the updated service IP allocations: serviceipallocations "" cannot be updated: the provided resource version does not match
!!! [0113 20:03:30] Timed out waiting for apiserver:  to answer at http://127.0.0.1:8080/healthz; tried 120 waiting 1 between each
!!! Error in ./hack/test-update-storage-objects.sh:48
  'return 1' exited with status 1
Call stack:
  1: ./hack/test-update-storage-objects.sh:48 startApiServer(...)
  2: ./hack/test-update-storage-objects.sh:123 main(...)
Exiting with status 1
```

karlkfi commented:
Apparently `github.com/docker/docker/pkg/symlink` is dual licensed... So I think we're going to have to add an exception. `fs.go` is bsd and `fs_test.go` is apache-2. So I think we have to go with BSD for the import.

ref: https://github.com/docker/docker/tree/master/pkg/symlink

karlkfi commented:
Added `github.com/docker/docker/pkg/symlink:spdxBSD3` exception

karlkfi commented:
@lavalamp Finally got this wrapped up and properly automated. It doesn't do verification that the exceptions cover all unknown licenses, but it does at least validate that the licenses file has been changed when someone changes the Godeps.

lavalamp commented on hack/verify-all.sh:
```diff
@@ -60,7 +60,7 @@ if $SILENT ; then
 
 # remove protobuf until it is part of direct generation
-EXCLUDE="verify-godeps.sh verify-generated-protobuf.sh"
+EXCLUDE="verify-godeps.sh verify-godep-licenses.sh verify-generated-protobuf.sh"
```
Why exclude? Is it slow?

karlkfi replied:
It's not super fast, but I figured if we weren't checking godeps there wasn't much point in checking something that depends on the godeps...

lavalamp commented on hooks/pre-commit:
```diff
@@ -28,6 +28,16 @@ if [[ "${#files[@]}" -ne 0 ]]; then
   fi
   echo "${reset}"
+
+  echo -ne "Check if Godep licesnses are up to date..."
```
nit: spelling

karlkfi replied:
oops! catch it next time we update the file, I guess.

lavalamp commented:
Wow, it looks like you went through a lot of effort-- my comments are optional

LGTM
